[clang][AArch64] Move initialization of ptrauth-* function attrs

atrosinenko opened the pull request:
Move the initialization of ptrauth-* function attributes near the initialization of branch protection attributes. The semantics of these groups of attributes partially overlaps, so handle both groups in getDefaultFunctionAttributes() and setTargetAttributes() functions to prevent getting them out of sync. This fixes C++ TLS wrappers.

atrosinenko committed:
[clang][AArch64] Move initialization of ptrauth-* function attrs

Move the initialization of ptrauth-* function attributes near the
initialization of branch protection attributes. The semantics of these
groups of attributes partially overlaps, so initialize them both in
getTrivialDefaultFunctionAttributes() function to prevent getting them
out of sync. This fixes C++ TLS wrappers.

DanielKristofKiss reviewed (approved):
LGTM.
maybe worth to review the attribute handling for the synthetic function's attributes in llvm too.

ojhunt reviewed (approved):
Looks good, I may start working on moving the preferences into the targetinfo unless you're working on that already

atrosinenko commented:
It seems that when I rebased this PR from an older version of `main` branch, some test failures came up. I'm investigating these now. An example is [clang/test/CodeGen/ptrauth-function-lvalue-cast.c](https://github.com/llvm/llvm-project/blob/00c5cd8a7a3f7e363d8f0d172ea88af91612c7fa/clang/test/CodeGen/ptrauth-function-lvalue-cast.c#L12), where an attribute was added at call site:
```llvm
  call void %0() #1 [ "ptrauth"(i32 0, i64 0) ]
;                ^^
```

> maybe worth to review the attribute handling for the synthetic function's attributes in llvm too.

@DanielKristofKiss As far as I know, at least one example of synthetic functions not protected with pac-ret is Function Multi Versioning resolvers. This could be fixed via a separate PR, though.

atrosinenko committed:
Fix tests: don't init on call site, update in setTargetAttributes

atrosinenko commented:
Fixed the tests in 85c21ecb922826ae8fabb4bc0c518933f19e0634:
* do not add attributes on the call site (something was probably wrong with the way I was running the tests, as this should fail before the rebase as well...)
* added `setPointerAuthFnAttributes` as a counterpart of `setBranchProtectionFnAttributes` (in [clang/test/CodeGenCXX/ptrauth.cpp](https://github.com/llvm/llvm-project/blob/00c5cd8a7a3f7e363d8f0d172ea88af91612c7fa/clang/test/CodeGenCXX/ptrauth.cpp#L21) `std::terminate()` is implemented, but by the time its implementation is emitted, `GetOrCreateLLVMFunction` was already called without a `Decl`, thus lazy-initializing a function, but without the attributes)

> I may start working on moving the preferences into the targetinfo unless you're working on that already

@ojhunt I'm not working on it, thank you!
